hopr-socks/src/main.rs

```diff
@@ -0,0 +1,203 @@
+use anyhow::Context;
+use fast_socks5::client::{Config as ClientConfig, Socks5Stream};
+use fast_socks5::server::{Authentication, Config as ServerConfig, SimpleUserPassword, Socks5Socket};
+use fast_socks5::Result;
+use log::{debug, error, info, warn};
+use std::future::Future;
+use std::sync::Arc;
+use structopt::StructOpt;
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    net::TcpListener,
+    task,
+};
+
+/// # How to use it:
+///
+/// ## Run in server mode:
+///
+/// $ hopr-socks --host 127.0.0.1 --port 1337 server no-auth
+/// $ hopr-socks --host 127.0.0.1 --port 1337 server password --username admin --password password
+///
+/// ## Run in client mode:
+///
+/// $ hopr-socks --host 127.0.0.1 --port 1337 client --target-host example.com no-auth
+///
+/// $ hopr-socks --host 127.0.0.1 --port 1337 client --target-host example.com password --username admin --password password
+///
+#[derive(Debug, StructOpt)]
+#[structopt(name = "hopr-socks", about = "A simple SOCKS5 server implementation.")]
+
+struct Opt {
+    /// Bind on address address. eg. `127.0.0.1`
+    #[structopt(long)]
+    host: String,
+
+    /// Bind on address address. eg. `1337`
+    #[structopt(long)]
+    port: String,
+
+    /// Choose running mode
+    #[structopt(subcommand, name = "mode")]
+    pub mode: RunModeOpt,
+}
+
+#[derive(StructOpt, Debug)]
+enum RunModeOpt {
+    Client {
+        /// Target address server (not the socks server)
+        #[structopt(short = "a", long)]
+        target_host: String,
+
+        /// Target port server (not the socks server)
+        #[structopt(short = "p", long, default_value = "80")]
+        target_port: u16,
+
+        /// Choose authentication type
+        #[structopt(subcommand, name = "auth")]
+        auth: AuthMode,
+    },
+    Server {
+        /// Request timeout
+        #[structopt(short = "t", long, default_value = "10")]
+        request_timeout: u64,
+
+        /// Choose authentication type
+        #[structopt(subcommand, name = "auth")]
+        auth: AuthMode,
+    },
+}
+
+/// Choose the authentication type
+#[derive(StructOpt, Debug)]
+enum AuthMode {
+    NoAuth,
+    Password {
+        #[structopt(short, long)]
+        username: String,
+
+        #[structopt(short, long)]
+        password: String,
+    },
+}
```

Comment:
You could import hopr-lib and use its exported definitions already at this point.

Reply:
To replace what ?

Reply:
Copy paste the config object with extensions from the hoprd (noauth/token).